src/event-manager/event-channel-client.ts

```diff
@@ -42,19 +44,34 @@ export default class EventChannelClient {
         this.handleMessage(message);
       });
       this._wsClient.on('error', (err) => {
-        reject(err);
+        console.error(err);
+        this.disconnect();
       });
       this._wsClient.on('open', () => {
         this._isConnected = true;
         resolve();
       });
-      // TODO(cshcomcom): Handle close connection (w/ ping-pong)
+      this._wsClient.on('ping', () => {
+        if (this._heartbeatTimeout) {
+          clearTimeout(this._heartbeatTimeout);
+        }
+        this._heartbeatTimeout = setTimeout(() => {
+          console.log(`Connection timeout! Terminate the connection. All event subscriptions are stopped.`);
+          this._wsClient.terminate();
+        }, connectionOption.heartbeatIntervalMs || (15000 + 1000)); // NOTE: This time must be longer than blockchain event handler heartbeat interval.
+      });
+      this._wsClient.on('close', () => {
+        this.disconnect();
+      });
     })
   }
 
   disconnect() {
     this._isConnected = false;
-    this._wsClient.close();
+    this._wsClient.terminate();
+    if (this._heartbeatTimeout) {
+      clearTimeout(this._heartbeatTimeout);
```

Comment:
minor) I think this still leaves this._heartbeatTimeout set as the timeout id (i.e. if (this._heartbeatTimeout) always evaluates to true) unless you set it to null, for instance.

Reply:
Oh thank you!